### PR TITLE
Fix race condition when fetching the same artifact

### PR DESCRIPTION
**Pull request by fdc-ntflx**
### Context

Explain context and other details for this pull request.

### Checklist

- [ ] `./gradlew build` compiles code correctly
- [ ] Added new tests where applicable
- [ ] `./gradlew test` passes all tests
- [ ] Extended README or added javadocs where applicable

### CODE REVIEW

File: mantis-server/mantis-server-agent/src/main/java/io/mantisrx/server/agent/HadoopFileSystemBlobStore.java [hunk 1]
```diff
@@ -30,22 +33,32 @@
...
     public File get(URI blobUrl) throws IOException {
         final Path src = new Path(blobUrl);
         final Path dest = new Path(getStorageLocation(blobUrl));
-        File destFile = new File(dest.toUri().getPath());
-        if (!destFile.exists()) {
-            fileSystem.copyToLocalFile(src, dest);
+        Lock lock = locks.get(dest);
+        try {
+            lock.lock();
+            log.info("Getting file with path {} from HadoopFileSystemBlobStore", dest);
```
Comment (sundargates) [inline mantis-server/mantis-server-agent/src/main/java/io/mantisrx/server/agent/HadoopFileSystemBlobStore.java:53]: nit: HadoopFileSystemBlobStore should already be part of the logger classname. No need to add it here too.

File: mantis-server/mantis-server-agent/src/main/java/io/mantisrx/server/agent/HadoopFileSystemBlobStore.java [hunk 2]
```diff
@@ -30,22 +33,32 @@
...
+            File destFile = new File(dest.toUri().getPath());
+            if (!destFile.exists()) {
+                fileSystem.copyToLocalFile(src, dest);
```
Comment (sundargates) [inline mantis-server/mantis-server-agent/src/main/java/io/mantisrx/server/agent/HadoopFileSystemBlobStore.java:56]: nit: shouldn't the lock be acquired only if the dest file does not exist?

File: mantis-server/mantis-server-agent/src/main/java/io/mantisrx/server/agent/HadoopFileSystemBlobStore.java [hunk 3]
```diff
@@ -30,22 +33,32 @@
...
 @RequiredArgsConstructor
+@Slf4j
 public class HadoopFileSystemBlobStore implements BlobStore {
 
     //  The file system in which blobs are stored. */
     private final FileSystem fileSystem;
 
     private final File localStoreDir;
 
+    final Striped<Lock> locks = Striped.lock(10);
```
Comment (sundargates) [inline mantis-server/mantis-server-agent/src/main/java/io/mantisrx/server/agent/HadoopFileSystemBlobStore.java:44]: nit: 1024 seems like a safer number?

**Review by sundargates — APPROVED** [review]

File: mantis-server/mantis-server-agent/src/main/java/io/mantisrx/server/agent/HadoopFileSystemBlobStore.java [hunk 4]
```diff
@@ -30,22 +33,32 @@
...
     @Override
     public File get(URI blobUrl) throws IOException {
         final Path src = new Path(blobUrl);
         final Path dest = new Path(getStorageLocation(blobUrl));
-        File destFile = new File(dest.toUri().getPath());
-        if (!destFile.exists()) {
-            fileSystem.copyToLocalFile(src, dest);
+        Lock lock = locks.get(dest);
+        try {
+            lock.lock();
```
Comment (Andyz26) [inline mantis-server/mantis-server-agent/src/main/java/io/mantisrx/server/agent/HadoopFileSystemBlobStore.java:52]: should lock.lock() be outside of the try?

**Pull request merged**